e2e/mandelbrot/steps_to_compile_mandelbrot.md

```diff
@@ -0,0 +1,930 @@
+    start_time = time.time()
+    func(*args)
+    end_time = time.time()
+    print(f"{name} : ", end_time - start_time, "sec")
+
+bench(mandelbrot_numpy, 'numpy', xmin, xmax, ymin, ymax, xn, yn, maxiter, horizon)
+```
+
+    numpy :  1.5263571739196777 sec
+
+
+# Torch.compile the iteration
+
+To improve the run time of the algorithm, we will need to apply several identical transformations. Let us work it one by one.
+
+
+## 1. Extract the inner iteration and only compile it
+
+The first thing to note is that torch dynamo agressively unrolls loops. Thus compiling the `mandelbrot` function directly would unroll the inner loop of `maxiter` iterations. This leads to very long compile times and extravagant memory consumption; on consumer grade machines compilation may even run out of memory. To avoid this, we extract the inner iteration and will only compile it, not the whole simulation:
+
+
+```python
+def step(n, C, Z, N, horizon):
+    I = np.abs(Z) < horizon
+    N[I] = n
+    Z[I] = Z[I]**2 + C[I]
+    return Z, N
+
+
+# note the additional argument, `step`
+def mandelbrot(xmin, xmax, ymin, ymax, xn, yn, maxiter, step, horizon=2.0):   
+    X = np.linspace(xmin, xmax, xn, dtype='float32')
+    Y = np.linspace(ymin, ymax, yn, dtype='float32')
+    C = X + Y[:, None]*1j
+    N = np.zeros(C.shape, dtype='int')
+    Z = np.zeros(C.shape, dtype='complex64')
+    for n in range(maxiter):
+        Z, N = step(n, C, Z, N, horizon)       
+    N[N == maxiter-1] = 0
+    return Z, N
+```
+
+
+```python
+# Set up the compilation
+
+import torch
+
+import torch._dynamo.config as cfg
+cfg.numpy_ndarray_as_tensor = True    # this will go when torch_np is upstreamed
+
+# make sure we all warnings from the compiler
+import warnings
+warnings.simplefilter('always')
+
+# Uncomment these two lines below to see the compilation logs and more detailed feedback from the compiler
+# (these will be relatively long and messy)
+import logging
+torch._logging.set_logs(dynamo=logging.WARNING)
+```
+
+The first attempt is to compile the `step` function as is. We will use the `fullgraph=True` argument to `torch.compile` so that the compilation fails on a graph break instead of falling back to eager mode.
+
+
+```python
+step_1 = torch.compile(fullgraph=True)(step)
+
+# run it once to warm up the JIT
+mandelbrot(xmin, xmax, ymin, ymax, xn, yn, maxiter, step_1, horizon)
+```
+
+
+    ---------------------------------------------------------------------------
+
+    DynamicOutputShapeException               Traceback (most recent call last)
+
+    File ~/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/_dynamo/utils.py:1353, in run_node(tracer, node, args, kwargs, nnmodule)
+       1352 if op == "call_function":
+    -> 1353     return node.target(*args, **kwargs)
+       1354 elif op == "call_method":
+
+
+    File ~/mambaforge/envs/torch_nightly/lib/python3.8/site-packages/torch/utils/_stats.py:20, in count.<locals>.wrapper(*args, **kwargs)
```

Comment:
Truncate this output with a `...` as it is always a pain to scroll through lol.
We won't show all this output in the tutorial as it's rather overwhelming.

Reply:
We'll mention it, but won't show it explicitly of course.